File: test/loan_policy_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:loan_app/core/loan/loan_policy.dart';

void main() {
  test('keeps the current 0.5% monthly flat-rate loan calculation', () {
    final quote = LoanPolicy.calculate(amount: 70000, termMonths: 4);

    expect(quote.principal, 70000);
    expect(quote.interestAmount, 1400);
    expect(quote.totalRepayment, 71400);
    expect(quote.monthlyPayment, 17850);
  });

  test('rejects loan products outside the approved rules', () {
    expect(
      () => LoanPolicy.calculate(amount: 69999, termMonths: 4),
      throwsArgumentError,
    );
    expect(
      () => LoanPolicy.calculate(amount: 70000, termMonths: 6),
      throwsArgumentError,
    );
  });
}

File: server/test/loan-policy.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';

import { calculateLoanQuote, createRepaymentSchedule, loanPolicy } from '../src/lib/loan-policy.js';

test('calculates the existing 0.5% flat-rate loan policy', () => {
  const quote = calculateLoanQuote(70000, 4);

  assert.deepEqual(quote, {
    principal: 70000,
    interestAmount: 1400,
    totalRepayment: 71400,
    monthlyPayment: 17850,
  });
});

test('creates an exact repayment schedule', () => {
  const schedule = createRepaymentSchedule(70000, 4, new Date('2026-01-01T00:00:00.000Z'));

  assert.equal(schedule.length, 4);
  assert.equal(schedule.reduce((sum, payment) => sum + payment.amountDue, 0), 71400);
  assert.equal(schedule[0].dueDate.toISOString(), '2026-02-01T00:00:00.000Z');
});

test('rejects an amount or term outside the approved product', () => {
  assert.throws(() => calculateLoanQuote(loanPolicy.minimumAmount - 1, 4));
  assert.throws(() => calculateLoanQuote(loanPolicy.minimumAmount, 6));
});
